Port OIDC (OpenID Connect) auth plugin from DSpace-CRIS

hardyoyo opened the pull request:
## References
* Along with https://github.com/DSpace/dspace-angular/pull/1470, this fixes #8087
* Requires https://github.com/DSpace/dspace-angular/pull/1470

## Description
Ports the OIDC (OpenID Connect) auth plugin from DSpace-CRIS

## Instructions for Reviewers
To fully test this plugin, you'll need to add it to your configuration, as you would any other [DSpace Authentication plugin](https://wiki.lyrasis.org/display/DSDOC7x/Authentication+Plugins). You will also need an OpenID Connect implementation. There are [Docker containers for you to use](https://www.keycloak.org/getting-started/getting-started-docker) or, if AWS is a resource available to you, you can use [AWS Cognito](https://aws.amazon.com/cognito/). 

If you need access to an OpenID Connect server, the NGLP staging authentication server is available, let me know and we can work out the details.

Draft documentation is here: https://wiki.lyrasis.org/x/dAPCDQ

List of changes in this PR:
* Adds OidcAuthentication.java, OidcAuthenticationBean.java, OidcClient.java, OidcRestController.java, OidcAuthenticationFilter.java
* Add integration tests in OidcAuthenticationRestControllerIT.java

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

Follow the draft docs: https://wiki.lyrasis.org/x/dAPCDQ and have an OIDC server you can use for testing. If you don't have one, contact me, I can probably help you.

## Checklist 
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

Commit:
[DSC-196] Porting of OpenID authentication provider

Commit:
[DSC-196] Fix tests

hardyoyo committed:
Add missing util: ThrowingSupplier

hardyoyo committed:
Ignoring jenv .java-version files

hardyoyo committed:
Add OrcidAuthenticationFiler

hardyoyo committed:
Suggested changes from Luca Giamminonni: remove ORCID AuthN

hardyoyo committed:
Remove unusued DSpace-CRIS services in core-services.xml

hardyoyo committed:
Disable DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES

- Makes OIDC authentication sturdier: implementations can and do differ
slightly, an unknown property should not prevent authentication

hardyoyo commented:
My apologies, we had a power outage today, and between that and other work priorities, I haven't yet been able to look at this PR. I will keep working on it as I can, however, my work is officially out for the holidays, and I should probably not work too much more on this, or risk side-eye from my family and colleagues.

I'm not sure why the build is failing, but I do see that the Shibboleth plugin has a similar but slightly different invocation for the line that's failing in this PR... I'm going to try using the line from the Shib plugin, and see if that gets things to compile.

hardyoyo commented:
Some things to consider in deliberating whether to accept this work: the code is from DSpace-CRIS, and we've deployed it on our DSpace pilot instance. It works fine. As soon as I figure out why the build fails with this rebased version on the current main branch, I'm sure it'll be mergable.

hardyoyo committed:
Revise call to create new DSpaceAuthentication

- Borrow the invocation of DSpaceAuthentication from the Shib plugin
- Also borrow the comment explaining it from the Shib plugin

hardyoyo committed:
Iron out some checkstyle wrinkles

hardyoyo commented:
This PR now passes tests. I will add docs to the description, I'll try to do so tomorrow.

hardyoyo committed:
Add missing isUsed override to OidcAuthentication code.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/OidcRestController.java:
```diff
@@ -0,0 +1,58 @@
+            redirectUrl = configurationService.getProperty("dspace.ui.url");
+        }
+        log.info("Redirecting to " + redirectUrl);
+        response.sendRedirect(redirectUrl);
```
LGTM is noted that this is a redirection of a potentially untrusted URL (as it comes from an external source), so it could be  security issue. https://lgtm.com/projects/g/DSpace/DSpace/rev/pr-19cb537c834fd1e3cc7e1caa2e89d4c1e92242ea

We had a similar issue with the Shibboleth implementation initially and fixed this by validating that the RedirectURL is "trusted" (and either corresponds to a configured DSpace UI or REST API).  See this method: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/security/ShibbolethLoginFilter.java#L110

You'll need to update this PR to do something similar. You could even move that logic to a shared Utils class if you wanted and update ShibbolethLoginFilter and this OidcRestController to call it there.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/OidcRestController.java:
```diff
@@ -0,0 +1,58 @@
+ */
+@RestController
+@RequestMapping(value = "/api/" + AuthnRest.CATEGORY + "/oidc")
+public class OidcRestController {
```
I'll just note here that the strategy of having both an `OidcRestController` **and** and `OidcAuthenticationFilter` might result in this "two auth cookies" bug that we hit with Shibboleth recently: https://github.com/DSpace/DSpace/pull/8032   Essentially, with Shibboleth, the controller & filter were both somehow creating an Auth Cookie, and fix was to remove the Controller entirely, and move all the logic into the Filter (see that PR as an example). 

While I cannot test OIDC to see if the same behavior will occur, I wanted to note here that this might be an issue we'll have to fix, as the OIDC code looks very similar to our old Shibboleth code where that error occurred.

hardyoyo replied:
If this is still a concern, I can try to tackle the change. I'll see if I can translate the changes in #8032 to this plugin

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/security/OidcAuthenticationFilter.java:
```diff
@@ -0,0 +1,50 @@
+ * @author Pasquale Cavallo (pasquale.cavallo at 4science dot it)
+ */
+
+public class OidcAuthenticationFilter extends StatelessLoginFilter {
```
This should be renamed `OidcLoginFilter` as it extends `StatelessLoginFilter` (and not `StatelessAuthenticationFilter`, which is a different class).  We don't want to imply this extends `StatelessAuthenticationFilter`.  Again, we had to fix this for Shibboleth as well in #8032

tdonohue reviewed (changes requested):
@hardyoyo : I tried building & running this backend today with a Keycloak Docker image (based loosely on https://gruchalski.com/posts/2020-09-03-keycloak-with-docker-compose/).  ~~While I can get Keycloak working via Docker, I **cannot** get the DSpace REST API backend to start up via this PR.  When starting the backend (via Tomcat) I always get this error:~~

~~I'm not exactly sure what's going on here, to be honest...but Spring/Tomcat cannot seem to find the `OidcAuthenticationBean` and everything crashes.~~  (FALSE ALARM!  My Docker environment had a different PR's code cached. After shutting everything down & rebuilding from scratch I'm able to get things running.  The below inline code comments are still valid however.)

I'll also point out that [LGTM has given an automated review](https://github.com/DSpace/DSpace/pull/8088#issuecomment-1006085568) and noted that this PR creates a new security alert.  We should get this fixed as well... I've noted inline below what it's complaining about.

Finally, I've found a few issues in the PR code here that are very similar to old Shibboleth behavior (and we fixed bugs in how Shibboleth was not acting correctly).  So, I think other classes in this PR will need some refactoring to fix those same bugs (but I cannot tell for certain until I can test this)... see more inline comments below.

tdonohue commented on dspace-api/src/main/java/org/dspace/authenticate/OidcAuthenticationBean.java:
```diff
@@ -0,0 +1,272 @@
+            return ePerson.canLogIn() ? logInEPerson(context, ePerson) : BAD_ARGS;
+        }
+
+        return canSelfRegister() ? registerNewEPerson(context, userInfo, email) : NO_SUCH_USER;
```
If canSelfRegister() returns false and no Eperson exists, this silently fails every single login.  Debugging that is a huge pain (I've been fighting this for a day now it seems).  We need to add a WARN or INFO logs here which can say something like "Self registration is currently disabled, but no EPerson could be found." or similar.

tdonohue commented on dspace-api/src/main/java/org/dspace/authenticate/OidcAuthentication.java:
```diff
@@ -0,0 +1,89 @@
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ *
+ */
+public class OidcAuthentication implements AuthenticationMethod {
```
I have to admit, this design of having an AuthenticationMethod which wraps a separate Bean which is also an AuthenticationMethod is very odd.  Could @LucaGiamminonni or someone explain why we'd want to do this?  Why do we need both an `OidcAuthentication` and `OidcAuthenticationBean`, both of which implement `AuthenticationMethod`?  Shouldn't we just use one like we do with all other Auth classes?

LucaGiamminonni replied:
Hi @tdonohue, The intent of this class is to be able to use the OidcAuthenticationBean class as a spring bean, with all the advantages that it entails: for example, I can take advantage of dependency injection instead of having to do the lookup to derive the dependencies and this also simplifies the test. The very odd thing for me is not the OidcAuthenticationBean class which is structured to be a classic spring bean but is that the others AuthenticationMethod are not beans of Spring. If someday we make all the other Spring AuthenticationMethod become beans (which I hope) the OidcAuthenticationBean class does not need to change.

tdonohue reviewed (changes requested):
@hardyoyo : So, I finally got this working in a local Docker setup!  As discussed on Slack privately, it was a pain, but my Docker setup is in #8109 which adds a few commits to this PR.

Once I got this working, it seems to work well. However, the pain of getting this working made me realize this PR is really hard to configure properly (and debug), which is why I have several inline suggestions on how to improve this PR's configs (most should be minor, I hope).

So, the good news is that this finally **works**.  But, I do think this PR needs some cleanup based on my code review....see inline comments below.

hardyoyo commented:
This is all great feedback, @tdonohue, thanks, I'll work on these suggested changes and get back to you soon.

hardyoyo commented on dspace/config/modules/authentication-oidc.cfg:
```diff
@@ -0,0 +1,42 @@
+#---------------------------------------------------------------#
+
+# The domain of the OpenID Connect server
+authentication-oidc.auth-server-domain =
```
you are correct, I'll remove it

hardyoyo committed:
Requested changes

- Add sane fall-back defaults for OIDC, where possible.
- Improve error logging for missing properties
- Include authentication-oidc.cfg in dspace.cfg
- Add configuration examples for OIDC to local.cfg-EXAMPLE
- Improve authentication-oidc.cfg with sane defaults and more comments

hardyoyo committed:
Requested change: Add logging for no such user and canSelfRegister=false

hardyoyo committed:
Fix checkstyle errors

hardyoyo committed:
Add validation to OIDC redirect, based on suggestion from LGTMbot, code copied from Shibboleth, at the suggestion of @tdonohue

hardyoyo committed:
Fix checkstyle error introduced by VScode's 'organize imports' suggestion

hardyoyo commented:
I want to give a shout-out to my UC colleague @ksclarke, who always answers my pleas for help, and offers helpful suggestions, such as using a map for the error message here:  https://github.com/DSpace/DSpace/pull/8088/commits/3a04b92c808207c592c67b78127bb242fcd33017#diff-3eb08922c76432a499d86cadd03ea1109da307c143198f33a3f7be971eb645ffR162-R172

hardyoyo committed:
Rename OidcAuthenticationFilter to OidcLoginFilter, as per request

hardyoyo commented:
> This pull request **introduces 1 alert** when merging [b0ce61d](https://github.com/DSpace/DSpace/commit/b0ce61d3fe933613af499beb591e1b0ee19d2639) into [7504269](https://github.com/DSpace/DSpace/commit/750426988bc6ff87d8ab7c42ebef31144b692597) - [view on LGTM.com](https://lgtm.com/projects/g/DSpace/DSpace/rev/pr-ddc02bf5bfea9a2c3a0cd0e6753ae4240a6569a5)
> 
> **new alerts:**
> 
> * 1 for URL redirection from remote source

For the record, LGTMbot, this has been addressed.

tdonohue reviewed (commented):
@hardyoyo : So, I'm having trouble getting this PR working again today in Docker. I *thought* I had everything working before, but now I'm back to hitting a 401 error on every login (last time I hit this, it was an HTTPS connection issue in Docker, but I thought I fixed it).  I pushed up an updated draft Docker PR #8136 with what I'm using.

So, while I wish I could approve this, I cannot seem to get it working again locally. I suspect it might be something in Docker again, but I've not had time to dig deeply yet.   That said, I have a few minor, inline comments based on an updated code review.

UPDATE TO ADD: I also agree with you that the LGTM issue seems fixed, so I think LGTM is reporting a false positive. We can tell LGTM to stop that by adding a comment at the end of line 70 (the redirect line) that says: `// lgtm [java/unvalidated-url-redirection]`.

hardyoyo committed:
Requested changes

- only warn about failure to find a matching ePerson if canSelfRegister
is false
- tell LGTMbot everything is fine with the redirect
- remove comments about nonexistent ORCID login code

tdonohue reviewed (approved):
👍 @hardyoyo Thanks for your hard work on this port, and thanks to @LucaGiamminonni & the 4Science team for the original code.

I've FINALLY been able to get this to work by using the OIDC test site https://www.phantauth.net/   Here's how to easily configure DSpace to use that test site in the `local.cfg`:
```
# Enable OIDC
plugin.sequence.org.dspace.authenticate.AuthenticationMethod = org.dspace.authenticate.OidcAuthentication

# Settings for OIDC authentication
# Based on instructions at https://www.phantauth.net/doc/integration
authentication-oidc.authorize-endpoint = https://phantauth.net/auth/authorize
authentication-oidc.token-endpoint = https://phantauth.net/auth/token
authentication-oidc.user-info-endpoint = https://phantauth.net/auth/userinfo

# Obtain a random client-id and client-secret via https://phantauth.net/client
# Place those values in these next two configs.
authentication-oidc.client-id = 
authentication-oidc.client-secret = 

# Because PhantAuth uses random users, you MUST ensure self registration is enabled
# (This is the default though)
# authentication-oidc.can-self-register = true
```

With those settings in place, I can login via PhantAuth, which creates a new DSpace EPerson and authenticates them.

All my other feedback has been addressed in this PR

tdonohue commented:
@LucaGiamminonni or @abollini : Could you give a final "sign off" on this port/migration (along with the UI PR at https://github.com/DSpace/dspace-angular/pull/1470)?  I would appreciate verification that 4Science doesn't see any problems accidentally introduced in this PR before merger (highly doubtful).   I'd like to merge this on Monday if possible.

hardyoyo committed:
Fix example OIDC line in local.cfg.EXAMPLE, wrong case!

LucaGiamminonni reviewed (approved):
Hi @hardyoyo, I tried this feature using google as provider and was able to log into the system. So for me the changes are ok and can be merged. Thank you

tdonohue commented:
Merging, as this is at +2.  Thanks again @hardyoyo and @LucaGiamminonni !
